### PR TITLE
ci: Track code coverage

### LINKED ISSUE #8615: Language provider binaries are not built for coverage tracking
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Minor enhancement idea. Currently Go-generated lang provider binaries (pulumi-language-python etc) are not being built with coverage support enabled, therefore do not participate in code coverage tracking scheme. In fact building them with coverage enabled breaks tests because of some inconsistencies in the CLI argument parsing scheme, the expectations of `pulumi` CLI as it constructs CLI arguments for these lang provider binaries are violated when the coverage-tracked builds include `--test` options.

If we mange to work this out, we will see the additional test coverage data that results from running code that is run within these lang provider binaries.

### PR DESCRIPTION
**Pull request by abhinav**
**Overview**

This re-enables tracking of code coverage.
For Go, there are two kinds of coverage at play:
unit test and integration test coverage.

Unit tests follow the usual pattern of running
`go test -cover -coverprofile=whatever.cov`.

For integration tests, we use the new integration test profiling support
[added in Go 1.20](https://go.dev/testing/coverage/).
In short, the way it works is:

    # Build a coverage instrumented binary:
    go build -cover

    # Set GOCOVERDIR to a directory and run the integration tests
    # that will invoke this coverage-instrumented binary.
    GOCOVERDIR=$(pwd)/coverage
    go test ./tests

    # $GOCOVERDIR will now be filled with coverage data
    # from every invocation of the coverage-instrumented binary.
    # Combine it into a single coverage file:
    go tool covdata textfmt -i=$(GOCOVERDIR) -o=out.cov

    # The resulting file can be uploaded to codecov as-is.

The above replaces the prior, partially working hacks we had in place
to get coverage-instrumented binaries with `go test -c`
and hijacking the TestMain.

**Notable changes**

- TestMain hijacking is deleted from the Pulumi CLI.
  We no longer need this to build coverage-instrumented binaries.
- ProgramTest no longer tracks or passes PULUMI_TEST_COVERAGE_PATH
  because the Pulumi binary no longer accepts a test.coverprofile flag.
  This information is now in the GOCOVERDIR environment variable.
- We add an `enable-coverage` parameter to the `ci-build-binaries`
  workflow to mirror some of the other workflows.
  It will produce coverage-instrumented binaries if this is true.
  These binaries are then used by `ci-run-test` which will set
  `GOCOVERDIR` and merge the coverage results from it.
- Coverage configuration no longer counts tests, testdata,
  and Protobuf-generated code against coverage.
- go-wrapper.sh:
  Because we're no longer relying on the `go test -c` hack,
  this no longer excludes Windows and language providers
  from coverage tracking.
- go-test.py and go-wrapper.sh will include pulumi-language-go and
  pulumi-language-nodejs in covered packages.

*Other changes*

- go-test.py:
  Fixed a bug where `args` parameters added for coverage were ignored.

Note that this change DOES NOT track coverage for calls made to Pulumi
packages by plugins downloaded from external sources,
e.g. provider plugins. Arguably, that's out of scope of coverage
trackcing for the Pulumi repository.

Resolves #8615, #11419

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13298** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13298" target="_blank">Graphite</a>  👈
    * **PR #13310** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13310" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/13298?utm_source=stack-comment).

File: scripts/go-test.py [hunk 1]
```diff
@@ -80,6 +94,7 @@ def heartbeat():
     args = ['go', 'test'] + args
 
 if not dryrun:
+    print("Running: " + ' '.join(args))
```
Comment (abhinav) [inline scripts/go-test.py:97]: Not a debugging leftover.
It's useful to know the exact call it's running.

**Comment by abhinav** [comment]
bors try

**Comment by abhinav** [comment]
There's some UI weirdness with there not being a real base result for codecov to compare against.
I believe the output should be stable once this lands.
I'm going to mark this as _not_ resolving #8629 until we know for sure.

**Comment by abhinav** [comment]
Also note that coverage is only tracked for PRs right now, including community PR tests run with `run-acceptance-tests`.
Merge builds run by bors have `enable-coverage` set to false. This can be changed if we want to track coverage at merge time as well.

**Comment by abhinav** [comment]
> ## try
> Build failed:
> 
> * [bors-ok](https://github.com/pulumi/pulumi/actions/runs/5405559425/jobs/9822143848)

It's green

**Review by justinvp — APPROVED** [review]
LGTM

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
Note that this is all non-blocking. Just informative for now.

**Pull request merged**